script/application.go

```diff
@@ -0,0 +1,198 @@
+	Problems  []error           `json:"-"`
+
+	Account     string    `json:"account"`
+	Project     Project   `json:"project"`
+	Applicant   Applicant `json:"applicant"`
+	CanContact  bool      `json:"can_contact"`
+	ApproverId  int       `json:"approver_id,omitempty"`
+	IssueNumber int       `json:"issue_number"`
+	CreatedAt   time.Time `json:"created_at"`
+}
+
+func (a *Application) Parse(issue github.Issue) {
+	a.validator = Validator{}
+
+	if strings.Contains(*issue.Title, "[project name]") {
+		a.validator.AddError("Application title", *issue.Title, "is missing project name")
+	}
+
+	a.sections = a.extractSections(*issue.Body)
+
+	if isTestingIssue() {
+		data, err := json.MarshalIndent(a.sections, "", "\t")
+		if err != nil {
+			log.Fatalf("Could not marshal Sections input data: %s", err.Error())
+		}
+
+		debugMessage("Parsed input data:", string(data))
+	}
+
+	a.CreatedAt = issue.CreatedAt.Time
+	a.IssueNumber = *issue.Number
+	a.Account = a.stringSection("Account URL", IsPresent, ParseAccountUrl)
+	a.boolSection("Non-commercial confirmation", IsPresent, ParseCheckbox, IsChecked)
+
+	a.Project.IsTeam = a.boolSection("Team application", ParseCheckbox)
+	a.Project.IsEvent = a.boolSection("Event application", ParseCheckbox)
+
+	isProject := !a.Project.IsTeam && !a.Project.IsEvent
+
+	a.Project.Name = a.stringSection("Project name", IsPresent, ParsePlainString)
+	a.Project.Description = a.stringSection("Short description", IsPresent, ParsePlainString)
+	a.Project.Contributors = a.intSection("Number of team members/core contributors", IsPresent, ParsePlainString)
+	a.Project.HomeUrl = a.stringSection("Homepage URL", IsPresent, IsUrl)
+	a.Project.RepoUrl = a.stringSection("Repository URL", IsUrl)
+	a.Project.LicenseType = a.stringSection("License type", When(isProject, IsPresent), ParsePlainString)
+	a.Project.LicenseUrl = a.stringSection("License URL", When(isProject, IsPresent), IsUrl)
+	a.boolSection("Age confirmation", When(isProject, IsPresent), ParseCheckbox, When(isProject, IsChecked))
+
+	a.Applicant.Name = a.stringSection("Name", IsPresent, ParsePlainString)
+	a.Applicant.Email = a.stringSection("Email", IsPresent, IsEmail)
+	a.Applicant.Role = a.stringSection("Project role", IsPresent)
+	a.Applicant.Id = *issue.User.ID
+
+	a.stringSection("Profile or website", IsUrl)
+	a.stringSection("Additional comments")
+
+	a.CanContact = a.boolSection("Can we contact you?", ParseCheckbox)
+
+	if isTestingIssue() {
+		debugMessage("Application data:", a.GetData())
+	}
+
+	for _, err := range a.validator.Errors {
+		a.Problems = append(a.Problems, fmt.Errorf(err.Error()))
+	}
+}
+
+func (a *Application) IsValid() bool {
+	return len(a.Problems) == 0
+}
+
+func (a *Application) GetData() string {
+	data, err := json.MarshalIndent(a, "", "\t")
+	if err != nil {
+		log.Fatalf("Could not marshal Application data: %s", err.Error())
+	}
+
+	return string(data)
+}
+
+func (a *Application) extractSections(body string) map[string]string {
```

Comment:
I think this function could either use a comment, or maybe restructure it to be a little more clear - I'm having trouble wrapping my head around the logic here. Granted, I haven't written code in a while so I'm certainly a little rusty. :P

One thing that maybe could help is a convention in go to check non-happy path scenarios first, and return/continue early. This can help reduce the level of indentation and keep the happy path flowing through from top to bottom, which might make it a little easier to follow!

Reply:
Have updated this to hopefully be a little easier to follow, with comments for context

Reply:
Thank you! This helps a lot! Super clear now ❤️